ofrak_components/ofrak_components/flash.py

```diff
@@ -0,0 +1,695 @@
+            )
+
+        while data_count < data_len:
+            # The rest of the blocks are data blocks
+            data_count += block_data_len
+            data_block_count += 1
+        return data_block_count
+
+    def iterate_through_all_blocks(
+        self, data_len: int, includes_oob: bool
+    ) -> Generator[Iterable[FlashField], None, int]:
+        count = 0
+        for c in self.get_block_formats():
+            num_blocks_of_type = 1
+            if c == self.data_block_format:
+                num_blocks_of_type = self.get_num_data_blocks(data_len, includes_oob)
+
+            for _ in range(0, num_blocks_of_type):
+                yield c
+                count += 1
+        return count
+
+    def get_total_oob_size(self, data_len: int) -> int:
+        total_oob_size = 0
+        for c in self.get_block_formats():
+            block_oob_size = self.get_oob_size_in_block(c)
+            if block_oob_size is not None:
+                num_blocks = 1
+                if c is self.data_block_format:
+                    num_blocks = self.get_num_data_blocks(data_len)
+                for _ in range(0, num_blocks):
+                    total_oob_size += block_oob_size
+        return total_oob_size
+
+    def get_total_field_size(self, data_len: int, field_type: FlashFieldType) -> int:
+        total_field_size = 0
+        for c in self.get_block_formats():
+            block_field_size = self.get_field_length_in_block(c, field_type)
+            if block_field_size is not None:
+                num_blocks = 1
+                if c == self.data_block_format:
+                    num_blocks = self.get_num_data_blocks(data_len)
+                for _ in range(0, num_blocks):
+                    total_field_size += block_field_size
+        return total_field_size
+
+
+#####################
+#    IDENTIFIER     #
+#####################
+class FlashEccIdentifier(Identifier[FlashConfig]):
+    """
+    Identify an ECC protected region by searching for the magic bytes
+    """
+
+    targets = (FlashResource,)
+
+    async def identify(self, resource: Resource, config=FlashConfig):
+        if config.ecc_config.ecc_magic is not None:
+            data = await resource.get_data()
+            if config.ecc_config.ecc_magic not in data:
+                raise IdentifierError("Flash magic bytes present but not found in resource")
+            resource.add_tag(FlashEccProtectedResource)
+
+
+#####################
+#     UNPACKERS     #
+#####################
+class FlashEccProtectedResourceUnpacker(Unpacker[FlashConfig]):
+    """
+    Unpack regions of flash protected by ECC
+    """
+
+    targets = (FlashEccProtectedResource,)
+    children = (
+        FlashEccResource,
+        FlashHeaderBlock,
+        FlashBlock,
+        FlashTailBlock,
+        FlashLogicalDataResource,
+        FlashLogicalEccResource,
+    )
+
+    async def unpack(self, resource: Resource, config: FlashConfig = FlashConfig):
```

Comment:
This function is pretty huge, and it seems like there are some discrete bits of functionality that could be factored out into their own functions. Doing so might make this code a bit more readable and maintainable.